libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/BytesWriteOperation.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.nio;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.function.BiConsumer;
+
+public class BytesWriteOperation implements WriteOperation {
+
+    private final NioSocketChannel channel;
+    private final BiConsumer<Void, Throwable> listener;
+    private final ByteBuffer[] buffers;
+    private final int[] offsets;
+    private final int length;
+    private int internalIndex;
+
+    public BytesWriteOperation(NioSocketChannel channel, ByteBuffer[] buffers, BiConsumer<Void, Throwable> listener) {
+        this.channel = channel;
+        this.listener = listener;
+        this.buffers = buffers;
+        this.offsets = new int[buffers.length];
+        int offset = 0;
+        for (int i = 0; i < buffers.length; i++) {
+            ByteBuffer buffer = buffers[i];
+            offsets[i] = offset;
+            offset += buffer.remaining();
+        }
+        length = offset;
+    }
+
+    @Override
+    public BiConsumer<Void, Throwable> getListener() {
+        return listener;
+    }
+
+    @Override
+    public NioSocketChannel getChannel() {
+        return channel;
+    }
+
+    public boolean isFullyFlushed() {
```

Comment:
can we assert that it is less or equal?

Reply:
I'm not sure what you mean here. I added an assertion in the `incrementIndex` method to ensure that we never increment past the length.

Reply:
Oh I understand. I added that assertion.

Reply:
++